internal/builder/builder.go

```diff
@@ -394,6 +367,94 @@ func (b *Builder) Save(logger logging.Logger, creatorMetadata CreatorMetadata) e
 
 // Helpers
 
+func addBuildpacks(logger logging.Logger, tmpDir string, image imgutil.Image, additionalBuildpacks []dist.Buildpack, bpLayers dist.BuildpackLayers) error {
+	type buildpackToAdd struct {
+		tarPath   string
+		diffID    string
+		buildpack dist.Buildpack
+	}
+
+	buildpacksToAdd := map[string]buildpackToAdd{}
+	for i, bp := range additionalBuildpacks {
+		// create buildpack directory
+		bpTmpDir := filepath.Join(tmpDir, strconv.Itoa(i))
+		if err := os.MkdirAll(bpTmpDir, os.ModePerm); err != nil {
+			return errors.Wrap(err, "creating buildpack temp dir")
+		}
+
+		// create tar file
+		bpLayerTar, err := dist.BuildpackToLayerTar(bpTmpDir, bp)
+		if err != nil {
+			return err
+		}
+
+		// generate diff id
+		diffID, err := dist.LayerDiffID(bpLayerTar)
+		if err != nil {
+			return errors.Wrapf(err,
+				"getting content hashes for buildpack %s",
+				style.Symbol(bp.Descriptor().Info.FullName()),
+			)
+		}
+
+		bpInfo := bp.Descriptor().Info
+		// check against builder layers
+		if existingBPInfo, ok := bpLayers[bpInfo.ID][bpInfo.Version]; ok {
+			if existingBPInfo.LayerDiffID == diffID.String() {
+				logger.Debugf("Buildpack %s already exists on builder with same contents, skipping...", style.Symbol(bpInfo.FullName()))
+				continue
+			}
+
+			logger.Debugf(
+				`buildpack %s already exists on builder and will be overwritten
+  - existing diffID: %s
+  - new diffID: %s`,
+				style.Symbol(bpInfo.FullName()),
+				style.Symbol(existingBPInfo.LayerDiffID),
+				style.Symbol(diffID.String()),
+			)
+		}
+
+		// check against other buildpacks to be added
+		if otherAdditionalBP, ok := buildpacksToAdd[bp.Descriptor().Info.FullName()]; ok {
+			if otherAdditionalBP.diffID == diffID.String() {
+				logger.Debugf("Buildpack %s with same contents is already being added, skipping...", style.Symbol(bpInfo.FullName()))
+				continue
+			}
+
+			logger.Debugf(
+				`buildpack %s was previously defined with different contents
+  - previous diffID: %s
+  - using diffID: %s`,
+				style.Symbol(bpInfo.FullName()),
+				style.Symbol(otherAdditionalBP.diffID),
+				style.Symbol(diffID.String()),
+			)
+		}
```

Comment:
Is there some way we can collapse these two, or make the contents of the `if` funcs equivalent so it can be pulled out and reused? It makes the method much more intimidatingly long.

Reply:
Simplified the error message itself, but I couldn't see a clean way to combine the `if` statements here ☹️